cmake: boards: Fix issue with relative paths

nordicjm opened the pull request:
Fixes an issue with relative paths causing a cmake warning to be emitted.

Fixes #58812

yashi commented:
Hi,  Thanks for the fix.  Indeed it eliminates the warning. But with the fix, the build system still carries the converted path.  That means that the other parts of the build system will check the same location twice.

Also, in CMake, an absolute path and a real path is different.  Under non-windows, any path starting with `/` is a absolute path.  So `/../../` is a absolute path but not a real path.

Would it be better to have a function like

```cmake
function(sanitize_board_root outlist)
  set(newlist)
  foreach(item ${ARGN})
    file(REAL_PATH "${item}" item)
    list(APPEND newlist ${item})
  endforeach()
  list(REMOVE_DUPLICATES newlist)
  set(${outlist} ${newlist} PARENT_SCOPE)
endfunction()
```

And call it like `sanitize_board_root(BOARD_ROOT ${BOARD_ROOTA})` just before the `foreach()`?

I'm not good at naming a function, please forgive me about that.

nordicjm commented:
> So /../../ is a absolute path but not a real path.

Not sure I follow what you mean, that is an absolute path and there is no way it can be valid, so the build system should error out if that path is provided to it, if you wanted a relative path to go up 2 directories it would be `../../` not `/../..`

yashi commented:
Applogize not to be clear.

Your patch checks for absoluteness.  But when it's not, you converts to a real path.  That's an inconsistency right there.

If you just need an absolute path, there is `cmake_path([ABSOLUTE_PATH] ...)`, which doesn't check the actual file system but returns an absolute path.  Thus, it's fast.

I've used a real path because it returns a unique path and allows me to use `REMOVE_DUPLICATES`.  `file(REAL_PATH)`, in addition to symbolic resolution, makes `.` to `/path/to/dir/` instead of `/path/to/dir/.`.     Note that the last dot.

Despite the all things about, I feel like it's not the right place to fix this problem.  Is it right to use `zephyr_get` on `BOARD_ROOT`?  Why does it add both a relateive path and an absolute path?  If that's the what the function does, should we use it?  I don't know about `sbuild` usage, which is mentioned in `root.cmake`.   I'm a bit confused about it.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2268,6 +2269,22 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
       endif()
     endforeach()
 
+    if(SYSBUILD)
+      foreach(path ${${FILE_APPLICATION_ROOT}})
```
afaict this will adjust any relative path to absolute path when current Zephyr build is invoked by sysbuild.

What we should do instead is to use `zephyr_file()` from within sysbuild itself to adjust any relative paths given by the user as `-D` before writing sysbuild cache.

For example here: https://github.com/zephyrproject-rtos/zephyr/blob/a17fc4ff2b71cc833ae479c2aaf87d014fd87a5b/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L319
where any `*_ROOT` can be adjusted by `zephyr_file()` (where zephyr_file needs to adjust according to `APP_DIR` fot the common root.

The image specific relative root, such as `mcuboot_*_ROOT` can also be adjusted here, or postponed till image specific call to `zephyr_file`, similar to your current proposal. And then cache the image specific sysbuild var locally, similar to this:
https://github.com/zephyrproject-rtos/zephyr/blob/a17fc4ff2b71cc833ae479c2aaf87d014fd87a5b/cmake/modules/extensions.cmake#L2256-L2260

so that only cached and image specific sysbuild vars are updated.

Note, if doing it later, then you must be take out only the sysbuild image specific form of `<image>_<VAR>_ROOT`, which is not posiible today.

nordicjm replied:
Updated code

nordicjm committed:
cmake: boards: Fix issue with relative paths

Fixes an issue with relative paths both with and without using
sysbuild where they would not be updated properly or a warning
would previously be emitted.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>
Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2268,6 +2269,22 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
+        # For sysbuild images, allow rewriting paths to be relative from the
+        # application source directory.
+        if(NOT IS_ABSOLUTE ${path})
+          set(abs_path ${APPLICATION_SOURCE_DIR}/${path})
```
Updated code

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -182,6 +182,17 @@ function(ExternalZephyrProject_Add)
+    if(DEFINED CACHE{${ZBUILD_APPLICATION}_${type}_ROOT} AND NOT IS_ABSOLUTE $CACHE{${ZBUILD_APPLICATION}_${type}_ROOT})
+      set(rel_path $CACHE{${ZBUILD_APPLICATION}_${type}_ROOT})
+      cmake_path(ABSOLUTE_PATH rel_path BASE_DIRECTORY "${ZBUILD_SOURCE_DIR}" NORMALIZE OUTPUT_VARIABLE abs_path)
+      set(${ZBUILD_APPLICATION}_${type}_ROOT ${abs_path} CACHE PATH "" FORCE)
```
minor nit.

In case a user examines the CMakeCache, then knowing what changed the setting may be nice.
```suggestion
      set(${ZBUILD_APPLICATION}_${type}_ROOT ${abs_path} CACHE PATH "Sysbuild adjusted absolute path" FORCE)
```

Feel free to adjust any way you like, here and in other file.

nordicjm committed:
doc: release: 3.4: Add build system relative path fixes

Adds a note that some relative path issues in the build system
have been fixed.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@yashi could you give this a try?

yashi commented:
@nordicjm It's already merged but it works fine on my system.  Thanks!
